### PR TITLE
[DS][48/n] Allow AllPartitionsSubsets in TimeWindowPartitionMapping

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

As title -- I noticed that this caused an error upwards in the stack. There's no reason we can't map from an AllPartitionsSubset, we just need to convert it into a more explicit form first.

### CODE REVIEW

**Comment by OwenKephart** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22058?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#21975** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21975?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#21989](https://github.com/dagster-io/dagster/pull/21989) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21989?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#22048](https://github.com/dagster-io/dagster/pull/22048) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22048?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#22058** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22058?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#21974** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21974?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21973** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21973?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21972** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21972?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21971** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21971?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21853** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21853?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#21854](https://github.com/dagster-io/dagster/pull/21854) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21854?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @OwenKephart and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster/dagster/_core/definitions/time_window_partition_mapping.py [hunk 1]
```diff
@@ -155,10 +174,29 @@ def get_downstream_partitions_for_partitions(
...
+                cast(TimeWindowPartitionsDefinition, upstream_partitions_def),
+                "upstream_partitions_def",
+                TimeWindowPartitionsDefinition,
+            ),
+            to_partitions_def=check.inst_param(
+                cast(TimeWindowPartitionsDefinition, downstream_partitions_def),
+                "downstream_partitions_def",
+                TimeWindowPartitionsDefinition,
+            ),
+            from_partitions_subset=from_partitions_subset,
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/time_window_partition_mapping.py:199]: <i class='graphite__hidden'>[Re: lines 136 to 199]<br/><br/></i>

this seems a little copy-paste-y. anyway to consolidate?<span class='graphite__hidden'><br/><br/>See this comment inline on <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22058?utm_source=unchanged-line-comment">Graphite</a>.</span>

**Review by schrockn — CHANGES_REQUESTED** [review]
Makes sense. Let's see if we can dedup those codepaths a smidge.

**Comment by OwenKephart** [comment]
@schrockn updated

**Review by schrockn — APPROVED** [review]

**Comment by OwenKephart** [comment]
### Merge activity

* **Jun 4, 9:49 AM EDT**: @OwenKephart started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/22058).

**Pull request merged**